File: src/main/java/com/hosle/framework/libnetwork/RxHttpTask.java
package com.hosle.framework.libnetwork;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.hosle.framework.libnetwork.converter.stringable.StringableGsonConverterFactory;
import com.hosle.framework.libnetwork.rxretrofit.RtfServiceHandler;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import okhttp3.CookieJar;
import okhttp3.Dns;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import rx.Subscription;

/**
 * Created by tanjiahao on 16/9/27.
 * Original Project HomanNetwork
 */

public abstract class RxHttpTask<S> implements IRequestData, RtfServiceHandler<S> {

    private String HOST;
    private String BASEURL;

    private Retrofit retrofit;
    protected Subscription taskSubscription;//retrofit建立连接后返回的订阅者实例,用于取消

    protected Map<String, String> mUrlParams = new HashMap<String, String>();
    protected Map<String, String> mFormParams = new HashMap<String, String>();

    private static volatile OkHttpClient httpClient = null;

    private String TAG = null;

    public RxHttpTask() {
        initCommonParams();
    }

    public static OkHttpClient getHttpClient(CookieJar cookieJar) {
        if (null == httpClient)
            synchronized (RxHttpTask.class) {
                if (null == httpClient) {
                    httpClient = new CustomOkHttpClient()
                            .setCookieJar(cookieJar)
                            .createOkHttpClient();
                }
            }
        return httpClient;
    }



    public RxHttpTask createUrl(String baseUrl) {
        if (!TextUtils.isEmpty(baseUrl)) {
            BASEURL = baseUrl;
            Uri uri = Uri.parse(BASEURL);

            String scheme = uri.getScheme();

            HOST = uri.getScheme() + "://" + uri.getHost();

            int port = uri.getPort();

            if ("http".equals(scheme)) {

                if (-1 != port && 80 != port) {

                    HOST += (":" + port);

                }

            } else if ("https".equals(scheme)) {

                if (-1 != port && 443 != port) {

                    HOST += (":" + port);

                }

            }

        }
        return this;
    }

    public abstract CookieJar createCookieStore();

    public Retrofit createRetrofit() {

        retrofit = new Retrofit.Builder()
                .client(getHttpClient(createCookieStore()))
                .addConverterFactory(StringableGsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(HOST)
                .build();

        return retrofit;
    }


    public Retrofit createRetrofit(ResponseBodyStringListener listener) {

        retrofit = new Retrofit.Builder()
                .client(getHttpClient(createCookieStore()))
                .addConverterFactory(StringableGsonConverterFactory.create(listener))
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(HOST)
                .build();

        return retrofit;
    }

    @Override
    public S createService(String apiUrlString) {
        Type t = getClass().getGenericSuperclass();
        Type[] params = ((ParameterizedType) t).getActualTypeArguments();
        Class<S> cls = (Class<S>) params[0];
        S service = this.createUrl(apiUrlString)
                .createRetrofit()
                .create(cls);

        return service;
    }

    @Override
    public S createService(String apiString, ResponseBodyStringListener listener) {
        Type t = getClass().getGenericSuperclass();
        Type[] params = ((ParameterizedType) t).getActualTypeArguments();
        Class<S> cls = (Class<S>) params[0];
        S service = this.createUrl(apiString)
                .createRetrofit(listener)
                .create(cls);

        return service;
    }

    private void initCommonParams() {
//        addURLParams("from", "na-android");
//        String timeStamp = System.currentTimeMillis() + "";
//        addURLParams("request_time", timeStamp);

    }

    public void addURLParams(String name, String value) {
        if(!TextUtils.isEmpty(value)){
            value=value.trim();
        }
        mUrlParams.put(name, value);
    }

    public void addFormParams(String name, String value) {
        if (!TextUtils.isEmpty(name)) {
            if(!TextUtils.isEmpty(value)){
                value=value.trim();
            }
            mFormParams.put(name, TextUtils.isEmpty(value) ? "" : value);
        }
    }


    public Map<String, String> getFormParams() {
        return mFormParams;
    }

    public Map<String, String> getUrlParams() {
        return mUrlParams;
    }

    public Subscription getTaskSubscription() {
        return taskSubscription;
    }

    public void setTaskSubscription(Subscription taskSubscription) {
        this.taskSubscription = taskSubscription;
    }

    public String getTAG() {
        return TAG;
    }

    public void setTAG(String TAG) {
        this.TAG = TAG;
    }
}

File: src/main/java/com/hosle/framework/libnetwork/v2/RxHttpTask.kt
package com.hosle.framework.libnetwork.v2

import android.net.Uri
import android.text.TextUtils
import android.util.Log
import com.google.gson.internal.LinkedTreeMap
import com.google.gson.reflect.TypeToken
import com.hosle.framework.libnetwork.*
import com.hosle.framework.libnetwork.converter.stringable.StringableGsonConverterFactory
import com.hosle.framework.libnetwork.rxretrofit.CommonSubscriber
import com.hosle.framework.libnetwork.rxretrofit.OnSubscriberListener
import okhttp3.CookieJar
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Func1
import rx.schedulers.Schedulers
import java.lang.reflect.ParameterizedType
import java.util.HashMap
import kotlin.reflect.KClass

/**
 * Created by tanjiahao on 2018/6/21
 * Original Project HomanNetwork
 *
 */
abstract class RxHttpTask<S,M> : IRequestData<M> {

    open var url: String = ""
    private var HOST: String? = null

    private var urlParams: MutableMap<String, Any> = HashMap()
    private var formParams: MutableMap<String, Any> = HashMap()

    private var TAG: String? = null

    private var subscription:Subscription? = null

    override fun doGetData(onStartCallback: (() -> Unit)?, onFinishCallback: (() -> Unit)?, onSuccessCallback: ((M) -> Unit)?, onFailureCallback: ((Throwable) -> Unit)?): Subscription {

        val subscription = doRequestObservable()
                .subscribe(LambdaSubscriber<M>(onStartCallback, onFinishCallback, onSuccessCallback, onFailureCallback))
        this.subscription = subscription
        return subscription
    }

    override fun doPostData(onStartCallback: (() -> Unit)?, onFinishCallback: (() -> Unit)?, onSuccessCallback: ((M) -> Unit)?, onFailureCallback: ((Throwable) -> Unit)?): Subscription {
        val subscription = doRequestObservable()
                .subscribe(LambdaSubscriber<M>(onStartCallback, onFinishCallback, onSuccessCallback, onFailureCallback))
        this.subscription = subscription
        return subscription
    }

    override fun doGetData(listener: OnSubscriberListener<M>): Subscription {
        val subscription = doRequestObservable()
                .subscribe(CommonSubscriber<M>(listener))
        this.subscription = subscription
        return subscription
    }

    override fun doPostData(listener: OnSubscriberListener<M>): Subscription {
        val subscription =  doRequestObservable()
                .subscribe(CommonSubscriber<M>(listener))
        this.subscription = subscription
        return subscription
    }

    override fun cancel() {
        subscription?.unsubscribe()
    }

    fun createService(apiUrlString: String): S {

        val t = javaClass.genericSuperclass
        val params = (t as ParameterizedType).actualTypeArguments
        val cls = params[0] as Class<S>

        return createUrl(apiUrlString)
                .createRetrofit()
                .create(cls)
    }

    fun createService(apiString: String, listener: ResponseBodyStringListener): S {
        val t = javaClass.genericSuperclass
        val params = (t as ParameterizedType).actualTypeArguments
        val cls = params[0] as Class<S>

        return createUrl(apiString)
                .createRetrofit(listener)
                .create(cls)
    }


    fun getPath(): String {
        return url.substring(HOST!!.length + 1)
    }

    private fun createUrl(baseUrl: String): RxHttpTask<*,*> {
        if (!TextUtils.isEmpty(baseUrl)) {
            val uri = Uri.parse(baseUrl)

            val scheme = uri.scheme

            HOST = uri.scheme + "://" + uri.host

            val port = uri.port

            if ("http" == scheme) {

                if (-1 != port && 80 != port) {

                    HOST += ":" + port

                }

            } else if ("https" == scheme) {

                if (-1 != port && 443 != port) {

                    HOST += ":" + port

                }

            }

        }
        return this
    }

    abstract fun createCookieStore(): CookieJar

    private fun createRetrofit(): Retrofit {

        return Retrofit.Builder()
                .client(HttpClientSingleton.getInstance(createCookieStore()))
                .addConverterFactory(StringableGsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(HOST!!)
                .build()
    }


    private fun createRetrofit(listener: ResponseBodyStringListener): Retrofit {

        return Retrofit.Builder()
                .client(HttpClientSingleton.getInstance(createCookieStore()))
                .addConverterFactory(StringableGsonConverterFactory.create(listener))
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(HOST!!)
                .build()
    }


    abstract fun resetCommonFormParams(): Map<String, Any>?
    abstract fun resetCommonUrlParams():Map<String,Any>?

    fun addURLParam(name: String, value: Any) {
        urlParams[name] = value
    }

    fun addFormParam(name: String, value: Any) {
        if (!TextUtils.isEmpty(name)) {
            formParams[name] = value
        }
    }

    fun addURLParams(params: Map<String, Any>) {
        urlParams.putAll(params)
    }

    fun addFormParams(params: Map<String, Any>) {
        formParams.putAll(params)
    }

    private val transValueType: (Map.Entry<*, *>) -> String = { entry -> entry.value.toString() }

    fun getAllFormParams(): Map<String, String> {
        return HashMap<String, String>().apply {
            putAll(resetCommonFormParams()?.mapValues(transValueType) ?: emptyMap())
            putAll(formParams.mapValues(transValueType))
        }
    }

    fun getAllUrlParams(): Map<String, String> {
        return HashMap<String, String>().apply {
            putAll(resetCommonUrlParams()?.mapValues(transValueType) ?: emptyMap())
            putAll(urlParams.mapValues(transValueType))
        }
    }

    fun getTAG(): String? {
        return TAG
    }

    fun setTAG(TAG: String) {
        this.TAG = TAG
    }
}